### ISSUE TITLE
Missing Γ_C factor in backward extraction makes all 32-fold 50-bit SIS bounds exceed q

### ISSUE THREAD

**Issue by quangvdao** [osdnk/pikku#2]
### Background and motivation

I found this while reproducing the implementation and communication figures for *PikkuFold: Efficient Folding in a Few Kilobytes*. At repository commit [`175b2ecc14a7ade06373f0498653f55bf78ebef2`](https://github.com/osdnk/pikku/commit/175b2ecc14a7ade06373f0498653f55bf78ebef2), I reproduced the configured $m=2^{20}$ schemelet run and its reported in-protocol communication:

```text
projection_kb = 2.257
sumcheck_rounds_kb = 1.005
debatching_kb = 2.265
total_communication_kb = 5.527
output_check = verified
```

While tracing the commitment ranks back through `estimates.ipynb`, I noticed that the notebook's recurrence for extracting through 32 sequential folds does not match Theorem 1. The symbolic theorem includes the challenge expansion factor $\Gamma_{\mathcal C}$, while the concrete notebook drops it from every backward step.

I independently recomputed the 50-bit and 64-bit schedules below using the committed constants and the pinned Lattice Estimator at commit `53da5982597709ba0fdf94ea37a84d822310fd84`.

### Summary

PikkuFold Theorem 1 gives the one-fold extracted-accumulator bound

```math
\beta_{\mathrm{acc}}'
=
\beta_{\mathrm{out}}'
+
k\Gamma_{\mathcal C}\beta_{\mathrm{in}}'.
```

This follows directly from reconstructing the preceding accumulator as

```math
w_{\mathrm{acc}}^*
=
u-\sum_{i=1}^{k}z_iw_i^*
```

and applying

```math
\|z_iw_i^*\|
\le
\|z_i\|_{\mathrm{op}}\|w_i^*\|
\le
\Gamma_{\mathcal C}\beta_{\mathrm{in}}'.
```

The concrete notebook instead adds only

```math
k\beta_{\mathrm{in}}'
```

for each earlier fold.

With the paper's implemented parameters

```math
k=2,\qquad
\Gamma_{\mathcal C}=8.357,\qquad
T=32,
```

restoring the missing factor makes the SIS-break radius exceed the approximately 50-bit modulus for every advertised implementation size:

| Total $\mathbb Z_q$ coordinates | Correct $\log_2\beta_{\mathrm{SIS}}$ | $\beta_{\mathrm{SIS}}/q$ |
|---:|---:|---:|
| $2^{25}$ | 50.350 | 1.275 |
| $2^{27}$ | 51.350 | 2.549 |
| $2^{29}$ | 52.350 | 5.099 |

Once $\beta_{\mathrm{SIS}}\ge q$, the SIS-break relation is trivially satisfiable: for any coordinate $j$,

```math
x=q e_j\ne0,\qquad
Fx=0\bmod q,\qquad
\|x\|=q\le\beta_{\mathrm{SIS}}.
```

Thus the concrete knowledge-soundness reduction gives no SIS security for the advertised 32-fold 50-bit schedules, independently of the commitment rank.

I believe the paper's one-fold theorem is correct as stated. The defect is in the notebook's multi-fold parameter recurrence and therefore in the concrete parameters justified by that notebook.

### Affected versions

- The notebook, schemelet parameters and 32-fold rank selection were introduced in [`4a1c91ad2cf291dfe7f48607068fc555680aab15`](https://github.com/osdnk/pikku/commit/4a1c91ad2cf291dfe7f48607068fc555680aab15).
- The defect remains present at current HEAD [`175b2ecc14a7ade06373f0498653f55bf78ebef2`](https://github.com/osdnk/pikku/commit/175b2ecc14a7ade06373f0498653f55bf78ebef2), dated 2026-08-26.
- No release tags are currently visible.
- Paper checked: IACR ePrint 2026/1809, PDF SHA-256 `389180da07ac09bd6d5e427ad8a7402f8a251f68024c2cb3b7e6030c235f2198`.
- Lattice Estimator checked: commit `53da5982597709ba0fdf94ea37a84d822310fd84`.

### Relevant locations

At current HEAD:

- [`estimates.ipynb`](https://github.com/osdnk/pikku/blob/175b2ecc14a7ade06373f0498653f55bf78ebef2/estimates.ipynb), function `extracted_sis_bounds`, defines `gamma` but computes:

  ```python
  extraction_before_first_round = (
      beta_out_last_round
      + K * beta_in_extracted * (FOLDING_ROUNDS - 1)
  )
  ```

  It subsequently computes:

  ```python
  extracted_accumulator = (
      extraction_before_first_round
      + K * beta_in_extracted
  )
  ```

  Both backward-extraction additions omit `gamma`.

- The same notebook explicitly rejects a schedule when:

  ```python
  if beta_sis >= q:
      return None, None
  ```

  so `beta_sis < q` is already treated as a required security boundary; the faulty recurrence is what keeps the 50-bit schedules below that boundary.

- [`pikku-fold-schemelet/src/config.rs`, lines 3–10](https://github.com/osdnk/pikku/blob/175b2ecc14a7ade06373f0498653f55bf78ebef2/pikku-fold-schemelet/src/config.rs#L3-L10) states that the configured ranks account for 32 sequential folds at input coefficient bound $2^4$.

- [`pikku-fold-schemelet/src/config.rs`, lines 12–28](https://github.com/osdnk/pikku/blob/175b2ecc14a7ade06373f0498653f55bf78ebef2/pikku-fold-schemelet/src/config.rs#L12-L28) fixes $k=2$, $\Gamma_{\mathcal C}=8.357$, three projection layers and the other implementation constants.

- [`pikku-fold-schemelet/src/fold.rs`, lines 10–20](https://github.com/osdnk/pikku/blob/175b2ecc14a7ade06373f0498653f55bf78ebef2/pikku-fold-schemelet/src/fold.rs#L10-L20) samples the short fixed-weight challenges whose multiplication operator norm is bounded by $\Gamma_{\mathcal C}$.

- Theorem 1 of [ePrint 2026/1809](https://eprint.iacr.org/2026/1809) states $\beta_{\mathrm{acc}}' := \beta_{\mathrm{out}}' + k\Gamma_{\mathcal C}\beta_{\mathrm{in}}'$.

### Correct backward recurrence

Let $T$ be the number of sequential folds, starting from a zero or negligible accumulator. Define

```math
\beta_{\mathrm{in}}=B\sqrt M
```

and, from the three-layer JL extraction loss,

```math
\beta_{\mathrm{in}}'
=
\sqrt{k}\,\beta_{\mathrm{in}}
\left(\frac{u}{\ell}\right)^3.
```

For the committed constants,

```math
\ell=\sqrt{27.37},\qquad
u=\sqrt{343.2},
```

so

```math
\left(\frac{u}{\ell}\right)^3
=44.4026328828.
```

The honest output after $T$ folds has bound

```math
B_T
=
T k\Gamma_{\mathcal C}\beta_{\mathrm{in}}.
```

To apply the extractor to the first fold, begin with the final output and rewind folds $T,T-1,\ldots,2$. These are $T-1$ backward transitions. Each transition adds

```math
k\Gamma_{\mathcal C}\beta_{\mathrm{in}}'.
```

Therefore the output-witness bound supplied to the first-fold extractor is

```math
B_{\mathrm{back}}
=
T k\Gamma_{\mathcal C}\beta_{\mathrm{in}}
+
(T-1)k\Gamma_{\mathcal C}\beta_{\mathrm{in}}'.
```

For $T=32$, this is

```math
B_{\mathrm{back}}
=
32k\Gamma_{\mathcal C}\beta_{\mathrm{in}}
+
31k\Gamma_{\mathcal C}\beta_{\mathrm{in}}'.
```

The number 31 is not a protocol constant; it is simply $T-1$. After extracting the first fold, reconstructing the initial accumulator incurs one final addition of $k\Gamma_{\mathcal C}\beta_{\mathrm{in}}'$.

Theorem 1 then defines

```math
\rho_{\mathrm{bind}}
=(2\Gamma_{\mathcal C})^k,
```

```math
\beta_{\mathrm{bind}}
=(k+1)B_{\mathrm{back}}(2\Gamma_{\mathcal C})^k,
```

and

```math
\beta_{\mathrm{SIS}}
=
\max\{
8\Gamma_{\mathcal C}B_{\mathrm{back}},
2\beta_{\mathrm{bind}}\rho_{\mathrm{bind}}
\}.
```

Equivalently,

```math
\beta_{\mathrm{SIS}}
=
\max\{
8\Gamma_{\mathcal C}B_{\mathrm{back}},
2(k+1)B_{\mathrm{back}}(2\Gamma_{\mathcal C})^{2k}
\}.
```

For $k=2$, the accumulator-collision branch dominates.

### Reproduction: 50-bit implementation parameters

Use:

```python
from math import sqrt, log2

q = 2**50 - 2687
k = 2
gamma = 8.357
folds = 32
input_bound = 2**4

ell = sqrt(27.37)
u = sqrt(343.2)
jl_loss = (u / ell)**3

for log_M in (25, 27, 29):
    M = 2**log_M

    beta_in = input_bound * sqrt(M)
    beta_in_extracted = sqrt(k) * beta_in * jl_loss

    beta_out_last = folds * k * gamma * beta_in
    beta_back = (
        beta_out_last
        + (folds - 1) * k * gamma * beta_in_extracted
    )

    beta_bind = (k + 1) * beta_back * (2 * gamma)**k
    rho_bind = (2 * gamma)**k

    beta_sis = max(
        8 * gamma * beta_back,
        2 * beta_bind * rho_bind,
    )

    print(log_M, log2(beta_sis), beta_sis / q)
```

This prints, up to rounding:

```text
25  50.3501808647  1.2747204237
27  51.3501808647  2.5494408475
29  52.3501808647  5.0988816949
```

For comparison, executing the current source recurrence—using the present nonzero JL extraction loss but omitting `gamma`—gives:

| $M$ | Current-source $\log_2\beta_{\mathrm{SIS}}$ | Correct $\log_2\beta_{\mathrm{SIS}}$ | Increase in $B_{\mathrm{back}}$ |
|---:|---:|---:|---:|
| $2^{25}$ | 47.449 | 50.350 | 7.468× |
| $2^{27}$ | 48.449 | 51.350 | 7.468× |
| $2^{29}$ | 49.449 | 52.350 | 7.468× |

The notebook's checked-in outputs are stale and show slightly different values, including a zero JL extraction loss. The source recurrence above is the relevant comparison; regenerating the notebook is part of the remediation.

### Necessary reset horizons at the 50-bit modulus

Holding all current constants fixed, the largest number of folds satisfying only the necessary condition

```math
\beta_{\mathrm{SIS}}<q
```

is:

| $M$ | Largest $T$ with $\beta_{\mathrm{SIS}}<q$ |
|---:|---:|
| $2^{25}$ | 25 |
| $2^{27}$ | 13 |
| $2^{29}$ | 7 |

These are not secure reset intervals by themselves. They only avoid the immediately trivial vector $q e_j$. The Lattice Estimator must still show the targeted computational security with the selected rank, so the actual permitted interval may be shorter.

### Reproduction: corrected 64-bit ranks

The theoretical estimates use

```math
q=2^{64}-59,\qquad B=2^{10}.
```

With the corrected recurrence, $\beta_{\mathrm{SIS}}<q$ for all nine Table 5 schedules, so these remain meaningful SIS instances. However, rerunning the notebook's pinned classical SIS estimator changes most minimum ranks:

| $M$ | $\varphi$ | Current rank | Correct $\log_2\beta_{\mathrm{SIS}}$ | Correct minimum rank | Classical $\log_2(\mathrm{rop})$ at that rank |
|---:|---:|---:|---:|---:|---:|
| $2^{25}$ | 64 | 30 | 58.924 | 34 | 132.50 |
| $2^{25}$ | 128 | 14 | 56.350 | 16 | 136.63 |
| $2^{25}$ | 256 | 7 | 56.042 | 8 | 138.30 |
| $2^{27}$ | 64 | 31 | 59.924 | 35 | 131.69 |
| $2^{27}$ | 128 | 15 | 57.350 | 16 | 131.34 |
| $2^{27}$ | 256 | 8 | 57.042 | 8 | 133.01 |
| $2^{29}$ | 64 | 32 | 60.924 | 36 | 131.15 |
| $2^{29}$ | 128 | 15 | 58.350 | 17 | 135.29 |
| $2^{29}$ | 256 | 8 | 58.042 | 9 | 145.92 |

The target used by the notebook is 129 classical estimator bits.

Because the fold-only transcript does not depend on the commitment rank, the “Protocol” column of Table 5 is unchanged. The two fresh commitments do depend on the rank. Using the notebook's average 62-bit compact cost for a uniform scalar, the corrected fresh-commitment sizes are approximately:

| $M$ | $\varphi=64$ | $\varphi=128$ | $\varphi=256$ |
|---:|---:|---:|---:|
| $2^{25}$ | 32.94 KiB | 31.00 KiB | 31.00 KiB |
| $2^{27}$ | 33.91 KiB | 31.00 KiB | 31.00 KiB |
| $2^{29}$ | 34.88 KiB | 32.94 KiB | 34.88 KiB |

The corresponding published values are 27–31 KiB.

### Impact

Confirmed:

- The notebook's backward-extraction recurrence disagrees with Theorem 1.
- The missing factor changes the 32-fold bound by approximately 7.47× for the 50-bit implementation parameters.
- All three advertised 50-bit, 32-fold schedules have $\beta_{\mathrm{SIS}}\ge q$.
- The corresponding SIS-break relation contains the public trivial witness $q e_j$, regardless of the commitment rank.
- The theoretical 64-bit schedules remain nontrivial but require larger ranks under the notebook's own pinned estimator.
- The 5.527 KiB fold-only communication measurement is reproducible as a size statistic because it excludes fresh commitments and does not depend on $n$.

Not claimed:

- This report does not construct an accepting malicious PikkuFold transcript.
- It does not prove that the protocol is insecure under every possible alternative analysis.
- It does not identify a flaw in the symbolic one-fold theorem.
- It does not claim that the 5.527 KiB transcript-size measurement itself is arithmetically caused by the missing $\Gamma_{\mathcal C}$.

The demonstrated consequence is that the current 32-fold, 50-bit parameter claim is not established by the paper's own reduction: its alternative SIS-break relation is trivial. The measured schemelet may still be useful after choosing a sound fold horizon, modulus and commitment rank.

### Suggested remediation

1. Correct both backward-extraction additions in `extracted_sis_bounds`:

   ```python
   extraction_before_first_round = (
       beta_out_last_round
       + gamma * K * beta_in_extracted * (FOLDING_ROUNDS - 1)
   )

   extracted_accumulator = (
       extraction_before_first_round
       + gamma * K * beta_in_extracted
   )
   ```

2. Centralize the recurrence in one helper parameterized by the fold count $T$, and test at least $T=1$, $T=2$ and $T=32$. The tests should make the $T-1$ backward transitions explicit.

3. Keep the existing `beta_sis >= q` rejection, but turn it into a required parameter-generation invariant with a clear diagnostic. A successful estimator result should never be emitted for a trivial SIS-break radius.

4. Regenerate all 50-bit parameters. The current 32-fold schedules cannot be repaired by increasing the commitment rank. Possible repairs include:

   - reducing the number of folds between exact norm resets;
   - increasing $q$;
   - reducing $\Gamma_{\mathcal C}$;
   - tightening the JL extraction loss; or
   - changing the extraction/reset design.

   The simple $\beta_{\mathrm{SIS}}<q$ horizons of 25, 13 and 7 above are only upper bounds on candidate reset intervals; each candidate must still meet the target SIS cost.

5. Regenerate the 64-bit Table 5 ranks and fresh-commitment sizes. The corrected ranks from the pinned estimator are shown above and should be independently confirmed in the project's intended Sage environment.

6. Re-execute and save `estimates.ipynb` from a single pinned environment. The checked-in outputs currently disagree with the source code, including a stored zero JL extraction loss and inconsistent rank values.

7. Update the paper and artifact documentation to distinguish:

   - one-fold knowledge error;
   - the maximum validated fold horizon before reset;
   - the SIS cost at that horizon;
   - fold-only communication; and
   - fresh-commitment communication.

8. Add regression tests that compare the notebook recurrence directly against Theorem 1 and that fail the three current 50-bit, 32-fold schedules.

The separate folded-output `u64` norm-overflow issue comes from the locked RoKoko dependency and is already documented in [`lattice-arguments/rokoko#95`](https://github.com/lattice-arguments/rokoko/issues/95). I would keep that issue separate from this parameter-estimation report.

I would be happy to provide the corrected notebook cell or split parameter regeneration from the small recurrence fix if that is easier to review.

**Comment by osdnk** [comment]
Hi,
thanks!

Fixed. I used avg norm analysis and almost recovered the loss :) 

cf. #3
